### PR TITLE
Transform key-value-document reference keys of the product sync.

### PR DESCRIPTION
**Pull request by ahmetoz**
#176

### CODE REVIEW

File: src/main/java/com/commercetools/project/sync/service/impl/ReferencesServiceImpl.java [hunk 1]
```diff
@@ -24,36 +34,40 @@ public ReferencesServiceImpl(@Nonnull final SphereClient ctpClient) {
...
-      @Nonnull final Set<String> productTypeIds) {
+      @Nonnull final Set<String> productTypeIds,
+      @Nonnull final Set<String> customObjectIds) {
 
     final Set<String> nonCachedProductIds = getNonCachedIds(productIds);
     final Set<String> nonCachedCategoryIds = getNonCachedIds(categoryIds);
     final Set<String> nonCachedProductTypeIds = getNonCachedIds(productTypeIds);
+    final Set<String> nonCachedCustomObjectIds = getNonCachedIds(customObjectIds);
 
     // if everything is cached, no need to make a request to CTP.
```
Comment (JudeNiroshan) [inline src/main/java/com/commercetools/project/sync/service/impl/ReferencesServiceImpl.java:66]: custom object keys are not cached yet. So, we cannot say `if everything is cached` anymore. Shouldn't we change the comment slightly?

**Review by JudeNiroshan — APPROVED** [review]

**Review by lojzatran — CHANGES_REQUESTED** [review]

File: src/main/java/com/commercetools/project/sync/service/impl/ReferencesServiceImpl.java [hunk 2]
```diff
@@ -98,4 +113,41 @@ private void cacheKeys(@Nullable final ResultingResourcesContainer resultsContai
...
+  }
+
+  private QueryPredicate<CustomObject<JsonNode>> buildCustomObjectIdsQueryPredicate(
+      @Nonnull final Set<String> customObjectIds) {
+    final List<String> idsSurroundedWithDoubleQuotes =
+        customObjectIds
+            .stream()
+            .filter(StringUtils::isNotBlank)
+            .map(customObjectId -> format("\"%s\"", customObjectId))
+            .collect(Collectors.toList());
```
[the remark this answers is not in the document]
Comment (ahmetoz) [reply src/main/java/com/commercetools/project/sync/service/impl/ReferencesServiceImpl.java:148]: I don't know why it's not working, I will check this later: https://github.com/commercetools/commercetools-project-sync/pull/176/commits/f7fd5712fc4a492216edeeb9827d5dd38cdfdc04

File: src/main/java/com/commercetools/project/sync/service/impl/ReferencesServiceImpl.java [hunk 3]
```diff
@@ -98,4 +113,41 @@ private void cacheKeys(@Nullable final ResultingResourcesContainer resultsContai
...
+
+    if (nonCachedCustomObjectIds.isEmpty()) {
+      return CompletableFuture.completedFuture(idToKey);
+    }
+
+    final Consumer<List<CustomObject<JsonNode>>> pageConsumer =
+        page ->
+            page.forEach(
+                resource ->
+                    idToKey.put(
```
[the remark this answers is not in the document]
Comment (ahmetoz) [reply src/main/java/com/commercetools/project/sync/service/impl/ReferencesServiceImpl.java]: https://github.com/commercetools/commercetools-project-sync/pull/176/commits/5b05d0384069f8b738689c892404b0f9ca0f619c

File: src/main/java/com/commercetools/project/sync/service/impl/ReferencesServiceImpl.java [hunk 4]
```diff
@@ -68,7 +82,8 @@ public ReferencesServiceImpl(@Nonnull final SphereClient ctpClient) {
             combinedResult -> {
               cacheKeys(combinedResult);
               return idToKey;
-            });
+            })
+        .thenCompose(ignored -> fetchCustomObjectKeys(nonCachedCustomObjectIds));
```
[the remark this answers is not in the document]
Comment (ahmetoz) [reply src/main/java/com/commercetools/project/sync/service/impl/ReferencesServiceImpl.java:86]: Also, this is not worked as expected, I will check the reasons later, also the method has some potential refactoring changes that I will check soon. For now added as a todo: https://github.com/commercetools/commercetools-project-sync/pull/176/commits/29fd8dfa385048d1fa89ee848d571a3b53e73144

File: src/main/java/com/commercetools/project/sync/service/impl/ReferencesServiceImpl.java [hunk 5]
```diff
@@ -1,59 +1,72 @@
...
     final Set<String> nonCachedCategoryIds = getNonCachedIds(categoryIds);
     final Set<String> nonCachedProductTypeIds = getNonCachedIds(productTypeIds);
+    final Set<String> nonCachedCustomObjectIds = getNonCachedIds(customObjectIds);
 
-    // if everything is cached, no need to make a request to CTP.
     if (nonCachedProductIds.isEmpty()
         && nonCachedCategoryIds.isEmpty()
         && nonCachedProductTypeIds.isEmpty()) {
-      return CompletableFuture.completedFuture(idToKey);
+      return fetchCustomObjectKeys(customObjectIds);
```
Comment (lojzatran) [inline src/main/java/com/commercetools/project/sync/service/impl/ReferencesServiceImpl.java:69]: Why you don't check in the if like other nonCachedIds but you fetch here?

**Review by lojzatran — APPROVED** [review]

File: src/main/java/com/commercetools/project/sync/service/impl/ReferencesServiceImpl.java (hunk 1 above)
Comment (ahmetoz) [reply src/main/java/com/commercetools/project/sync/service/impl/ReferencesServiceImpl.java]: https://github.com/commercetools/commercetools-project-sync/pull/176/commits/3bc33b61b25894111d0ef02ec8acb11ec957e1f7

File: src/main/java/com/commercetools/project/sync/service/impl/ReferencesServiceImpl.java (hunk 5 above)
Comment (ahmetoz) [reply src/main/java/com/commercetools/project/sync/service/impl/ReferencesServiceImpl.java:69]: two avoid code duplication,

same here those structure needs to be changed for better service implementation for this PR I want to focus the feature, I will try to update required issues soon, also the graph is not handling the pagination which is a potential issue (low prio but possible) https://github.com/commercetools/commercetools-project-sync/pull/176#discussion_r496672896

**Pull request merged**